x-pack/solutions/observability/plugins/apm/common/service_map/get_service_map_nodes.ts

```diff
@@ -0,0 +1,292 @@
+    }
+    if (!allNodesMap.has(destinationId)) {
+      allNodesMap.set(destinationId, { ...connection.destination, id: destinationId });
+    }
+  });
+
+  // Derive the rest of the map nodes from the connections and add the services
+  // from the services data query
+  services
+    .filter((service) => !FORBIDDEN_SERVICE_NAMES.includes(service[SERVICE_NAME]))
+    .forEach((service) => {
+      const id = service[SERVICE_NAME];
+      if (!allNodesMap.has(id)) {
+        allNodesMap.set(id, { ...service, id });
+      }
+    });
+
+  return allNodesMap;
+}
+
+function getAllServices(
+  allNodes: Map<string, ConnectionNode>,
+  destinationServices: ExitSpanDestination[]
+) {
+  const connectionFromDiscoveredServices: ServiceConnectionNode[] = [];
+
+  for (const { from, to } of destinationServices) {
+    const fromId = from.id;
+    const toId = to.id;
+
+    if (allNodes.has(fromId) && !allNodes.has(toId)) {
+      connectionFromDiscoveredServices.push({ ...to, id: toId });
+    }
+  }
+
+  // List of nodes that are services
+  const serviceNodes = new Map(
+    [...Array.from(allNodes.values()), ...connectionFromDiscoveredServices]
+      .filter((node): node is ServiceConnectionNode => !isExitSpan(node))
+      .map((node) => [node.id, node])
+  );
+
+  return serviceNodes;
+}
+
+function getExitSpans(allNodes: Map<string, ConnectionNode>) {
+  return new Map(
+    Array.from(allNodes.values()).reduce((acc, node) => {
+      if (!isExitSpan(node)) {
+        return acc;
+      }
+
+      const nodes = acc.get(node.id) ?? [];
+      nodes.push(node as ExternalConnectionNode);
+      acc.set(node.id, nodes);
+
+      return acc;
+    }, new Map<string, ExternalConnectionNode[]>())
+  );
+}
+
+function exitSpanDestinationsToMap(destinationServices: ExitSpanDestination[]) {
+  return destinationServices.reduce((acc, { from, to }) => {
+    acc.set(from.id, to);
+    return acc;
+  }, new Map<string, ServiceConnectionNode>());
+}
+
+function mapNodes({
+  allConnections,
+  anomalies,
+  nodes,
+  exitSpanDestinations,
+  services,
+}: {
+  allConnections: Connection[];
+  anomalies: ServiceAnomaliesResponse;
+  nodes: Map<string, ConnectionNode>;
+  services: Map<string, ServiceConnectionNode>;
+  exitSpanDestinations: ExitSpanDestination[];
+}) {
+  const exitSpanDestinationsMap = exitSpanDestinationsToMap(exitSpanDestinations);
+  const exitSpans = getExitSpans(nodes);
+  const anomaliesByServiceName = new Map(
+    anomalies.serviceAnomalies.map((item) => [item.serviceName, item])
+  );
+
+  const messagingSpanIds = new Set(
+    allConnections.filter(({ source }) => isExitSpan(source)).map(({ source }) => source.id)
+  );
+
+  // 1. Map external nodes to internal services
+  // 2. Collapse external nodes into one node based on span.destination.service.resource
+  // 3. Pick the first available span.type/span.subtype in an alphabetically sorted list
+  return Array.from(nodes.entries()).reduce((map, [id, node]) => {
```

Comment:
Instead of allocating an array just to do a reduce, why not make use of the iterator with a for of loop?

```ts
const mappedNodes = new Map<string, ConnectionNode>();

for (const [id, node] of nodes.entries()) {
  if (mappedNodes.has(id)) {
    continue;
  }
  // The rest of the owl...
}

return mappedNodes;
```

Reply:
That make sense.